src/clj/rems/middleware.clj

```diff
@@ -69,25 +69,29 @@
 (defn wrap-context [handler]
   (fn [request]
     (binding [context/*root-path* (calculate-root-path request)
-              context/*roles* (when context/*user*
-                                (set/union (roles/get-roles (getx-user-id))
-                                           (applications/get-all-application-roles (getx-user-id))))]
+              context/*roles* (cond-> #{}
+                                context/*user* (set/union (roles/get-roles (getx-user-id))
+                                                          (applications/get-all-application-roles (getx-user-id)))
+                                (:uses-api-key? request) (conj :api-key))]
       (with-mdc {:roles (str/join " " (sort context/*roles*))}
         (handler request)))))
 
 (defn wrap-role-headers [handler]
   (fn [request]
     (cond-> (handler request)
-      context/*roles* (header "x-rems-roles" (str/join " " (map name context/*roles*))))))
+      (not (empty? context/*roles*)) (header "x-rems-roles" (str/join " " (map name context/*roles*))))))
 
 (deftest test-wrap-role-headers
   (testing "no roles"
     (is (= {}
            (binding [context/*roles* nil]
+             ((wrap-role-headers identity) {}))))
+    (is (= {}
+           (binding [context/*roles* #{}]
              ((wrap-role-headers identity) {})))))
   (testing "one role"
     (is (= {:headers {"x-rems-roles" "foo"}}
-           (binding [context/*roles* [:foo]]
+           (binding [context/*roles* #{:foo}]
```

Comment:
nit: How about changing vector to a set in "multiple role" test case below, as well?

Reply:
With sets the order is undefined, so it would complicate asserting the headers... unless the headers are sorted. 💡 I guess I'll go do that. 👍